File: Core/src/main/java/jokatu/game/MultiInputStage.java
package jokatu.game;

import jokatu.game.event.GameEvent;
import jokatu.game.input.Input;
import jokatu.game.input.InputAcceptor;
import jokatu.game.player.Player;
import jokatu.game.stage.Stage;
import ophelia.collections.BaseCollection;
import ophelia.collections.UnmodifiableCollection;
import ophelia.event.observable.AbstractSynchronousObservable;
import ophelia.exceptions.StackedException;
import ophelia.exceptions.voidmaybe.VoidMaybe;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * A stage that can accept multiple types of {@link Input}s by using some {@link InputAcceptor}s
 */
public abstract class MultiInputStage extends AbstractSynchronousObservable<GameEvent> implements Stage<GameEvent> {

	private final List<InputAcceptor<? extends GameEvent>> inputAcceptors = new ArrayList<>();

	protected void addInputAcceptor(InputAcceptor<? extends GameEvent> inputAcceptor) {
		inputAcceptors.add(inputAcceptor);
		inputAcceptor.observe(this::fireEvent);
	}

	@NotNull
	@Override
	public UnmodifiableCollection<Class<? extends Input>> getAcceptedInputs() {
		Set<Class<? extends Input>> inputs = inputAcceptors.stream()
				.map(InputAcceptor::getAcceptedInputs)
				.flatMap(BaseCollection::stream)
				.collect(Collectors.toSet());
		return new UnmodifiableCollection<>(inputs);
	}

	@Override
	public final void accept(@NotNull Input input, @NotNull Player player) throws StackedException {
		VoidMaybe.mergeFailures(
				inputAcceptors.stream()
						.map(VoidMaybe.wrap(acceptor -> acceptor.accept(input, player)))
						.collect(Collectors.toList())
		).throwOnFailure();
	}
}

File: Core/src/main/kotlin/jokatu/game/MultiInputStage.kt
package jokatu.game

import jokatu.game.event.GameEvent
import jokatu.game.input.Input
import jokatu.game.input.InputAcceptor
import jokatu.game.player.Player
import jokatu.game.stage.Stage
import ophelia.event.observable.AbstractSynchronousObservable
import ophelia.exceptions.StackedException
import ophelia.exceptions.voidmaybe.VoidMaybe.wrap
import ophelia.exceptions.voidmaybe.VoidMaybeCollectors.merge
import java.util.*
import java.util.stream.Collectors
import kotlin.reflect.KClass

/**
 * A stage that can accept multiple types of [Input]s by using some [InputAcceptor]s
 */
abstract class MultiInputStage : AbstractSynchronousObservable<GameEvent>(), Stage<GameEvent> {

	private val inputAcceptors = ArrayList<InputAcceptor<out GameEvent>>()

	protected fun addInputAcceptor(inputAcceptor: InputAcceptor<out GameEvent>) {
		inputAcceptors.add(inputAcceptor)
		inputAcceptor.observe(::fireEvent)
	}

	override val acceptedInputs: Collection<KClass<out Input>>
		get() = inputAcceptors.stream()
				.map { it.acceptedInputs }
				.flatMap { it.stream() }
				.collect(Collectors.toSet())

	@Throws(StackedException::class)
	override fun accept(input: Input, player: Player) {
		inputAcceptors.stream()
				.map(wrap { acceptor -> acceptor.accept(input, player) })
				.collect(merge())
				.throwOnFailure()
	}
}
